### PR TITLE
Feat/crypto data hash predicate

### PR DESCRIPTION
**Pull request by ccamel**
Introduces the `crypto_data_hash/3` predicate, which closely looks like the one provided by [swiprolog](https://www.swi-prolog.org/pldoc/doc_for?object=crypto_data_hash/3), with current support for `MD5` (for convenience), `SHA-256` and `SHA-512`. This predicate offers more versatility (in terms of algorithm and encoding) compared to the ~~now-deprecated~~ **now-removed** `sha_hash/2` predicate.

Additionally, this PR addresses a bug related to parsing options for the empty list (`[]`), correcting the issue where the empty list was treated as a compound term, instead of an atom.

--
Note: The code for the `HashAlg` enum type has been generated with [abice/go-enum](https://github.com/abice/go-enum).

## Summary by CodeRabbit

- **New Features**
  - Introduced a new `crypto_data_hash/3` predicate for enhanced cryptographic hashing with multiple algorithms and encoding options.
  - Updated the `chain_id/1` predicate signature to `chain_id(?ID)` for improved clarity.
  - Added UTF-8 encoding support for term-to-byte conversions.

- **Improvements**
  - Deprecated the `sha_hash/2` predicate in favor of the more versatile `crypto_data_hash/3`.
  - Enhanced cryptographic verification functions to use the new `util.KeyAlg` type for specifying cryptographic algorithms.

- **Bug Fixes**
  - Improved error messages for invalid input types and unknown encoding options in term-to-byte conversions.
  - Fixed handling of empty lists in Prolog utility functions.

- **Documentation**
  - Updated documentation to reflect new predicates and deprecations.
  - Provided examples for using the new `crypto_data_hash/3` predicate.

- **Refactor**
  - Renamed `SHAHash` to `CryptoDataHash` to align with new functionality.
  - Changed function signatures to accommodate new hashing and encoding options.

- **Tests**
  - Added new test cases for `crypto_data_hash` and UTF-8 encoding.
  - Renamed test functions to reflect the updated cryptographic operations.

### CODE REVIEW

File: x/logic/predicate/crypto_test.go [hunk 1]
```diff
@@ -20,7 +20,7 @@ import (
 	"github.com/okp4/okp4d/x/logic/types"
 )
 
-func TestCryptoHash(t *testing.T) {
+func TestCrypto(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (ccamel) [reply x/logic/predicate/crypto_test.go]: Alright.

File: x/logic/predicate/crypto.go [hunk 2]
```diff
@@ -31,20 +32,94 @@
...
+				util.HashAlgNames()))
+		}
+		decodedData, err := TermToBytes(data, options, AtomUtf8, env)
+		if err != nil {
+			return engine.Error(fmt.Errorf("%s: failed to decode data: %w", functor, err))
+		}
+
+		result, err := util.Hash(algorithm, decodedData)
+		if err != nil {
+			engine.Error(fmt.Errorf("sha_hash/2: failed to hash data: %w", err))
```
[the remark this answers is not in the document]
Comment (ccamel) [reply x/logic/predicate/crypto.go]: That's true. But theres the variable `functor` for this.

File: x/logic/util/prolog_test.go [hunk 3]
```diff
@@ -23,6 +23,12 @@ func TestGetOption(t *testing.T) {
 				wantResult: nil,
 				wantError:  nil,
 			},
+			{
+				option:     engine.NewAtom("foo"),
+				options:    engine.List(),
+				wantResult: nil,
+				wantError:  nil,
+			},
```
[the remark this answers is not in the document]
Comment (ccamel) [reply x/logic/util/prolog_test.go:31]: No, not in this case. In prolog `Engine.List()` returns a `engine.Atom` which is `[]`, and therefore different from `nil`.

File: x/logic/util/crypto.go [hunk 4]
```diff
@@ -1,30 +1,54 @@
+//go:generate go-enum --names
 package util
 
 import (
 	"crypto/ecdsa"
 	"crypto/ed25519"
 	"crypto/elliptic"
+	"crypto/md5" //nolint:gosec
```
[the remark this answers is not in the document]
Comment (ccamel) [reply x/logic/util/crypto.go:8]: Of course it does. But this is a feature we want to provide via a prolog predicate. There's a warning about this in the description.

**Review by amimart — APPROVED** [review]
That's better 😉 

Maybe we can directly remove the `sha_hash/2` predicate as it is not used at this point?

**Comment by ccamel** [comment]
> Maybe we can directly remove the `sha_hash/2` predicate as it is not used at this point?

Oh yes sure. we can remove it. Want to take care of it 😛

**Comment by ccamel** [comment]
@coderabbitai pause

**Pull request merged**